### ISSUE TITLE
Memory leaks while rebinding vectors

### ISSUE THREAD

**Issue by Pawapek** [vrogier/ocilib#230]
Memory leaks while rebinding vectors

Hi,

I noticed that there is a problem with memory leaks when vector of values is rebinded. In my case I want to insert huge number of rows in parts(10,000 rows each).
This small code example illustrates the problem.

```cpp
int main()
{
  const int MAX_BIND_ARRAY_SIZE = 10;

  if (!ocilib::Environment::Initialized())
  {
    ocilib::Environment::Initialize(ocilib::Environment::Threaded);
  }

  ocilib::Connection connection;
  connection.Open(GetAlias(), GetUser(), GetPassword());

  ocilib::Statement stmt(connection);

  stmt.AllowRebinding(true);
  stmt.Prepare("INSERT INTO test_table(COLUMN1) VALUES (:1)");
  stmt.SetBindArraySize(MAX_BIND_ARRAY_SIZE);

  std::vector<unsigned int> values_vec;

  for (int i = 0; i < 2; ++i)
  {
    for (unsigned int j = 0; j < MAX_BIND_ARRAY_SIZE; ++j)
    {
      values_vec.push_back(j);
    }

    stmt.Bind(":1", values_vec, ocilib::BindInfo::In);
    stmt.ExecutePrepared();
  }
}
```

Executing this code results with valgrind detecting memory leaks:
==13162==    at 0x4C2A888: operator new[](unsigned long) (vg_replace_malloc.c:423)
==13162==    by 0x428EE2: ocilib::BindArray::BindArrayObject<unsigned int>::AllocData() (ocilib_impl.hpp:4686)
==13162==    by 0x42785B: ocilib::BindArray::BindArrayObject<unsigned int>::BindArrayObject(ocilib::Statement const&, std::string const&, std::vector<unsigned int, std::allocator<unsigned int> >&, bool, unsigned int, unsigned int) (ocilib_impl.hpp:4674)
==13162==    by 0x426088: void ocilib::BindArray::SetVector<unsigned int>(std::vector<unsigned int, std::allocator<unsigned int> >&, bool, unsigned int) (ocilib_impl.hpp:4629)
==13162==    by 0x424B94: void ocilib::Statement::BindVector1<int (OCI_Statement*, char const*, unsigned int*, unsigned int), unsigned int>(int (&)(OCI_Statement*, char const*, unsigned int*, unsigned int), std::string const&, std::vector<unsigned int, std::allocator<unsigned int> >&, ocilib::Enum<ocilib::BindInfo::BindDirectionValues>, ocilib::Enum<ocilib::BindInfo::VectorTypeValues>) (ocilib_impl.hpp:5304)
==13162==    by 0x423825: void ocilib::Statement::Bind<unsigned int>(std::string const&, std::vector<unsigned int, std::allocator<unsigned int> >&, ocilib::Enum<ocilib::BindInfo::BindDirectionValues>, ocilib::Enum<ocilib::BindInfo::VectorTypeValues>) (ocilib_impl.hpp:5594)

It looks like binded vector allocates memory(method **Statement::BindVector1** - creating BindArray object) but never gets deleted. I fixed this issue in my local application by adding "delete (*it);" to "BindsHolder::AddBindObject(BindObject *bindObject)" just before line "_bindObjects.erase(it);".

```cpp
@@ -5004,6 +5004,7 @@
         {
             if ((*it)->GetName() == bindObject->GetName())
             {
+                delete (*it);
                 _bindObjects.erase(it);
                 break;
             }
```

Please check if you can reproduce problem and if proposed fix is valid.

**Comment by vrogier** [comment]
Hi,

Thank for reporting issues.
I will investigate tomorrow.
If confirmed, a fix will made on the 4.7.0 branch.

Regards,

Vincent

**Comment by vrogier** [comment]
Proposed fix indeed eliminates the memory leak as the previous bind object was not freed.
But rather than systematically freeing and  recreating the bond object, I will check if resuing the same bind object is not too problematic to implement.

Regards,

Vincent
